filecoin-proofs/src/fr32_reader.rs

```diff
@@ -160,146 +108,44 @@ impl<R: io::Read> io::Read for Fr32Reader<R> {
-        debug_assert!(self.available() >= 1);
-        self.pos += 1;
-        res
-    }
-
-    #[cfg(target_endian = "little")]
-    pub fn read_u8_range(&mut self, len: u64) -> u8 {
-        use bitintr::Bextr;
-        debug_assert!(self.available() >= len);
-        let res = self.data.bextr(self.pos, len) as u8;
-        self.pos += len;
-        res
-    }
-
-    #[cfg(target_endian = "little")]
-    pub fn read_u8(&mut self) -> u8 {
-        use bitintr::Bextr;
-        debug_assert!(self.available() >= 8);
-        let res = self.data.bextr(self.pos, 8) as u8;
-        self.pos += 8;
-        res
-    }
-
-    #[cfg(target_endian = "little")]
-    pub fn read_u16(&mut self) -> u16 {
-        debug_assert!(self.available() >= 16);
-
-        use bitintr::Bextr;
-        let res = self.data.bextr(self.pos, 16) as u16;
-        self.pos += 16;
-        res
-    }
+        let mut read = 0;
 
-    #[cfg(target_endian = "little")]
-    pub fn read_u16_into(&mut self, target: &mut [u8]) {
-        assert!(target.len() >= 2);
+        while read < num_bytes {
+            // load new block
+            if self.to_process == 0 {
+                let bytes_read = self.fill_buffer()?;
 
-        let value = self.read_u16().to_le_bytes();
-        target[0] = value[0];
-        target[1] = value[1];
-    }
+                // read all data, no new data in the buffer
+                if bytes_read == 0 {
+                    self.done = true;
+                    break;
+                }
 
-    #[cfg(target_endian = "little")]
-    pub fn read_u32(&mut self) -> u32 {
-        debug_assert!(self.available() >= 32);
+                self.process_block();
+                self.to_process += div_ceil(bytes_read * 8, 254);
+                self.out_offset = 0;
+            }
 
-        use bitintr::Bextr;
-        let res = self.data.bextr(self.pos, 32) as u32;
-        self.pos += 32;
-        res
-    }
+            // write out result
+            let available_bytes = self.to_process * (256 / 8);
 
-    #[cfg(target_endian = "little")]
-    pub fn read_u32_into(&mut self, target: &mut [u8]) {
-        assert!(target.len() >= 4);
-        let value = self.read_u32().to_le_bytes();
-        target[0] = value[0];
-        target[1] = value[1];
-        target[2] = value[2];
-        target[3] = value[3];
-    }
+            let start = read;
+            let end = std::cmp::min(start + available_bytes, num_bytes);
+            let len = end - start;
 
-    pub fn read_u64(&mut self) -> u64 {
-        debug_assert!(self.available() >= 64);
+            let out_start = self.out_offset;
+            let out_end = out_start + len;
 
-        self.pos += 64;
-        self.data
-    }
+            target[start..end]
+                .copy_from_slice(&self.out_buffer.as_byte_slice()[out_start..out_end]);
+            read += len;
+            self.out_offset += len;
+            self.to_process -= div_ceil(len * 8, 256);
```

Comment:
`div_ceil(len, 32)` might be clearer. But this raises a question, if `to_process` is the number of blocks, and one block is 4 `Fr`, shouldn't this be `div_ceil(len, 128)`?

I can believe the actual algorithm is doing the right thing here, but I'm confused trying to line up all the names and descriptions throughout this whole changeset.

I think maybe 'block' in this context is being used to mean 'one `Fr`s worth of bytes' — but it seems like in the definition of `NUM_U128S_PER_BLOCK` (where the value is 8), 'block' is being used to refer to 128 bytes = 4 `Fr32`s.

Am I missing something, or can some rectification of names help?

Reply:
to process is in number of Frs sorry, which is 254 bits in, 256bits out